age/keysource.go

```diff
@@ -284,11 +287,105 @@ func (key *MasterKey) loadIdentities() (ParsedIdentities, error) {
 
 	var identities ParsedIdentities
 	for n, r := range readers {
-		ids, err := age.ParseIdentities(r)
-		if err != nil {
-			return nil, fmt.Errorf("failed to parse '%s' age identities: %w", n, err)
+
+		b := bufio.NewReader(r)
+		p, _ := b.Peek(14) // length of "age-encryption" and "-----BEGIN AGE"
+		peeked := string(p)
+
+		switch {
+		// An age encrypted file, plain or armored.
+		case peeked == "age-encryption" || peeked == "-----BEGIN AGE":
+			var r io.Reader = b
+			if peeked == "-----BEGIN AGE" {
+				r = armor.NewReader(r)
+			}
+			const privateKeySizeLimit = 1 << 24 // 16 MiB
+			contents, err := io.ReadAll(io.LimitReader(r, privateKeySizeLimit))
+			if err != nil {
+				return nil, fmt.Errorf("failed to read '%s': %w", n, err)
+			}
+			if len(contents) == privateKeySizeLimit {
+				return nil, fmt.Errorf("failed to read '%s': file too long", n)
+			}
+			IncorrectPassphrase := func() {
+				conn, err := gpgagent.NewConn()
+				if err != nil {
+					return
+				}
+				defer func(conn *gpgagent.Conn) {
+					if err := conn.Close(); err != nil {
+						log.Errorf("failed to close connection with gpg-agent: %s", err)
+					}
+				}(conn)
+				err = conn.RemoveFromCache(n)
+				if err != nil {
+					log.Warnf("gpg-agent remove cache request errored: %s", err)
+					return
+				}
+			}
+			ids := []age.Identity{&EncryptedIdentity{
+				Contents: contents,
+				Passphrase: func() (string, error) {
+					conn, err := gpgagent.NewConn()
+					if err != nil {
+						fmt.Fprintf(os.Stderr, "Enter passphrase for identity '%s':", n)
+
+						var pass string
+						if term.IsTerminal(int(os.Stdout.Fd())) {
+							p, err = term.ReadPassword(int(os.Stdout.Fd()))
+							if err == nil {
+								pass = string(p)
+							}
+						} else {
+							reader := bufio.NewReader(os.Stdin)
+							pass, err = reader.ReadString('\n')
+							if err == io.EOF {
+								err = nil
+							}
+						}
+						if err != nil {
+							return "", fmt.Errorf("could not read passphrase: %v", err)
+						}
+
+						fmt.Fprintln(os.Stderr)
+
+						return pass, nil
```

Comment:
#1692 adds `tui.go` in `age/` with some code vendored from age, namely the `readPassphrase` function (signature `func readPassphrase(prompt string) ([]byte, error)`). With that, I think this can be replaced by:
```suggestion
						passphrase, err := readPassphrase("Enter passphrase for identity '%s':")
						if err != nil {
							return "", err
						}
						return string(passphrase), nil
```